В этом домашнем задании мы будем работать с набором данных о пассажирах парохода "Титаник", потерпевшего крушение в 1912 году (источник: соревнование [Titanic: Machine Learning from Disaster](https://www.kaggle.com/c/titanic) на платформе [Kaggle](https://www.kaggle.com)).

# Базовый уровень

*Для получения положительной оценки нужно правильно выполнить не менее 4 заданий из 6*

## Задание 1.1
Загрузите таблицу из файла `data.csv` в переменную `df` типа `pd.DataFrame`. В качестве индекса следует использовать колонку `PassengerId`.

In [1]:
import pandas as pd

In [5]:
df = pd.read_csv('data.csv', sep=';', index_col='PassengerId')
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
862                 0       2   
863                 1       1   
864                 0       3   
865                 0       2   
866                 1       2   
867                 1       2   
868                 0       1   
869                 0       3   
870                 1       3   
871                 0       3   
872                 1       1   
873                 0       1   
874                 0       3   
875                 1       2   
876                 1       3   
877                 0       3   
878                 0       3   
879                 0       3   
880                 1       1   
881                 1       2   
882                 0       3   
883                 0       3   
884                 0       2   
885                 0       3   
886                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17                                        Rice, Master. Eugene    male   2.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
23           

## Задание 1.2
Сколько строк и столбцов в таблице?

In [6]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 11 columns):
Survived    891 non-null int64
Pclass      891 non-null int64
Name        891 non-null object
Sex         891 non-null object
Age         714 non-null float64
SibSp       891 non-null int64
Parch       891 non-null int64
Ticket      891 non-null object
Fare        891 non-null float64
Cabin       204 non-null object
Embarked    889 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 207.3 KB


Строки

In [111]:
df.index.max()

891

In [77]:
df.Name.count()

891

In [86]:
df.count(1).count()

891

Столбцы

In [78]:
df.count(axis=0).count()

11

строк - 891, столбцов - 11

## Задание 1.3
Выведите на экран последние три значения столбца `Age`(возраст).

In [35]:
df.Age.iloc[-3:]

PassengerId
889     NaN
890    26.0
891    32.0
Name: Age, dtype: float64

In [45]:
df.iloc[-3:,4]

PassengerId
889     NaN
890    26.0
891    32.0
Name: Age, dtype: float64

## Задание 1.4

Какова доля выживших (поле `Survived`)?

In [89]:
df.Survived.loc[df.Survived ==1].count()

342

In [90]:
df.Survived.count()


891

In [93]:
df.Survived.loc[df.Survived ==1].count()/df.Survived.count()

0.38383838383838381

## Задание 1.5
Выведите на экран таблицу, отсортированную по имени по алфавиту (поле `Name`).

In [94]:
df.sort_values('Name')

Survived  Pclass  \
PassengerId                     
846                 0       3   
747                 0       3   
280                 1       3   
309                 0       2   
875                 1       2   
366                 0       3   
402                 0       3   
41                  0       3   
856                 1       3   
208                 1       3   
811                 0       3   
841                 0       3   
211                 0       3   
785                 0       3   
731                 1       1   
5                   0       3   
306                 1       1   
298                 0       1   
499                 0       1   
835                 0       3   
193                 1       3   
461                 1       1   
851                 0       3   
814                 0       3   
120                 0       3   
69                  1       3   
542                 0       3   
543                 0       3   
14                  0       3   
147                 1       3   
...               ...     ...   
407                 0       3   
378                 0       1   
372                 0       3   
674                 1       2   
649                 0       3   
156                 0       1   
18                  1       2   
305                 0       3   
736                 0       3   
352                 0       1   
492                 0       3   
426                 0       3   
56                  1       1   
556                 0       1   
621                 0       3   
831                 1       3   
326                 1       1   
204                 0       3   
355                 0       3   
496                 0       3   
200                 0       2   
112                 0       3   
241                 0       3   
423                 0       3   
560                 1       3   
287                 1       3   
283                 0       3   
362                 0       2   
154                 0       3   
869                 0       3   

                                                        Name     Sex    Age  \
PassengerId                                                                   
846                                      Abbing, Mr. Anthony    male  42.00   
747                              Abbott, Mr. Rossmore Edward    male  16.00   
280                         Abbott, Mrs. Stanton (Rosa Hunt)  female  35.00   
309                                      Abelson, Mr. Samuel    male  30.00   
875                    Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.00   
366                           Adahl, Mr. Mauritz Nils Martin    male  30.00   
402                                          Adams, Mr. John    male  26.00   
41            Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.00   
856                               Aks, Mrs. Sam (Leah Rosen)  female  18.00   
208                              Albimona, Mr. Nassef Cassem    male  26.00   
811                                   Alexander, Mr. William    male  26.00   
841                              Alhomaki, Mr. Ilmari Rudolf    male  20.00   
211                                           Ali, Mr. Ahmed    male  24.00   
785                                         Ali, Mr. William    male  25.00   
731                            Allen, Miss. Elisabeth Walton  female  29.00   
5                                   Allen, Mr. William Henry    male  35.00   
306                           Allison, Master. Hudson Trevor    male   0.92   
298                             Allison, Miss. Helen Loraine  female   2.00   
499          Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.00   
835                                   Allum, Mr. Owen George    male  18.00   
193          Andersen-Jensen, Miss. Carla Christine Nielsine  female  19.00   
461                                      Anderson, Mr. Harry    male  48.00   
851                  Andersson, Maste

## Задание 1.6
Каковы доли женщин и мужчин среди пассажиров (поле `Sex`)?

Мужчины

In [105]:
df.Sex.loc[df.Sex =='male'].count()/df.Sex.count()

0.6475869809203143

Женщины

In [106]:
1 - df.Sex.loc[df.Sex =='male'].count()/df.Sex.count()

0.3524130190796857

# Продвинутый уровень

*Для получения положительной оценки нужно правильно выполнить не менее 3 заданий из 5*

## Задание 2.1
Каков процент женщин и мужчин среди пассажиров? (Hint: *процент – это доля, умноженная на 100.*)

Женщины

In [133]:
(df.Sex.loc[df.Sex =='female'].count()/df.Sex.count())*100

35.241301907968577

Мужчины

In [102]:
(df.Sex.loc[df.Sex =='male'].count()/df.Sex.count())*100

64.758698092031437

## Задание 2.2
Какова доля совершеннолетних пассажиров (18 лет и старше)?

In [132]:
df.Age.loc[df.Age >=18].count()/df.Age.count()

0.84173669467787116

## Задание 2.3
Замените коды пассажирских классов (поле `Pclass`) их названиями (1 = Upper, 2 = Middle, 3 = Lower). Результат выведите на экран.

In [113]:
 df.replace({'Pclass' : {1 : 'Upper', 2 : 'Middle',3:'Lower'}})

Survived  Pclass  \
PassengerId                     
1                   0   Lower   
2                   1   Upper   
3                   1   Lower   
4                   1   Upper   
5                   0   Lower   
6                   0   Lower   
7                   0   Upper   
8                   0   Lower   
9                   1   Lower   
10                  1  Middle   
11                  1   Lower   
12                  1   Upper   
13                  0   Lower   
14                  0   Lower   
15                  0   Lower   
16                  1  Middle   
17                  0   Lower   
18                  1  Middle   
19                  0   Lower   
20                  1   Lower   
21                  0  Middle   
22                  1  Middle   
23                  1   Lower   
24                  1   Upper   
25                  0   Lower   
26                  1   Lower   
27                  0   Lower   
28                  0   Upper   
29                  1   Lower   
30                  0   Lower   
...               ...     ...   
862                 0  Middle   
863                 1   Upper   
864                 0   Lower   
865                 0  Middle   
866                 1  Middle   
867                 1  Middle   
868                 0   Upper   
869                 0   Lower   
870                 1   Lower   
871                 0   Lower   
872                 1   Upper   
873                 0   Upper   
874                 0   Lower   
875                 1  Middle   
876                 1   Lower   
877                 0   Lower   
878                 0   Lower   
879                 0   Lower   
880                 1   Upper   
881                 1  Middle   
882                 0   Lower   
883                 0   Lower   
884                 0  Middle   
885                 0   Lower   
886                 0   Lower   
887                 0  Middle   
888                 1   Upper   
889                 0   Lower   
890                 1   Upper   
891                 0   Lower   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17                                        Rice, Master. Eugene    male   2.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
23           

## Задача 2.4
Какова доля выживших среди женщин? Среди мужчин?

Все , кто выжил

In [128]:
 df_Survived =df.loc[df.Survived ==1]

Количество выживших женщин к общему числу

In [130]:
(df_Survived.Sex.loc[df.Sex =='female'].count()/df.Sex.count())

0.26150392817059481

Количество выживших мужчин к общему числу

In [131]:
(df_Survived.Sex.loc[df.Sex =='male'].count()/df.Sex.count())

0.122334455667789

## Задача 2.5
Сколько всего пропущенных значений в таблице?

In [156]:
import numpy as np

Отдельно по столбцам

In [236]:
df.fillna('NULL').Cabin.loc[df.fillna('NULL').Cabin=='NULL'].count()

687

In [185]:
df.fillna('NULL').Age.loc[df.fillna('NULL').Age=='NULL'].count()

177

In [238]:
df.fillna('NULL').Embarked.loc[df.fillna('NULL').Embarked=='NULL'].count()

2

иным способом сразу по всем столбцам

In [233]:
df.loc[ 0:,'Cabin'].count()

204

In [234]:
df.index.max()-df.loc[ 0:,'Cabin'].count()

687

In [237]:
list_col=df.columns
sum_null_count=0
for i in list_col:
    if df.loc[ 0:,i].count()<df.index.max():
        sum_null_count+=df.index.max()-df.loc[ 0:,i].count()
print (sum_null_count)

866


# Сложный уровень

*Для получения положительной оценки нужно правильно выполнить не менее 2 заданий из 3*

## Задание 3.1
Добавьте в таблицу поле `Surname` – фамилию пассажира.

In [200]:
def surname_from_name(name):
    name_list = name.split(',')
    surname = name_list[0]
        #surname=i[0]
    return surname
surname_from_name('Dooley, Mr. Patrick')

'Dooley'

In [202]:
df.Name.apply(surname_from_name)

PassengerId
1               Braund
2              Cumings
3            Heikkinen
4             Futrelle
5                Allen
6                Moran
7             McCarthy
8              Palsson
9              Johnson
10              Nasser
11           Sandstrom
12             Bonnell
13         Saundercock
14           Andersson
15             Vestrom
16             Hewlett
17                Rice
18            Williams
19       Vander Planke
20          Masselmani
21              Fynney
22             Beesley
23             McGowan
24              Sloper
25             Palsson
26             Asplund
27                Emir
28             Fortune
29             O'Dwyer
30            Todoroff
            ...       
862              Giles
863              Swift
864               Sage
865               Gill
866            Bystrom
867       Duran y More
868           Roebling
869      van Melkebeke
870            Johnson
871             Balkic
872           Beckwith
873           Carlsson

In [203]:
df['Surname'] = df.Name.apply(surname_from_name)

In [204]:
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
862                 0       2   
863                 1       1   
864                 0       3   
865                 0       2   
866                 1       2   
867                 1       2   
868                 0       1   
869                 0       3   
870                 1       3   
871                 0       3   
872                 1       1   
873                 0       1   
874                 0       3   
875                 1       2   
876                 1       3   
877                 0       3   
878                 0       3   
879                 0       3   
880                 1       1   
881                 1       2   
882                 0       3   
883                 0       3   
884                 0       2   
885                 0       3   
886                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17                                        Rice, Master. Eugene    male   2.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
23           

## Задание 3.2
Найдите всех пассажиров с аномальной стоимостью билета (поле `Fare`). Будем считать стоимость аномальной, если она отличается от среднего более, чем на 5 стандартных отклонений. Выведите соответствующие строки таблицы на экран.

In [208]:
df.Fare.std()

49.693428597180898

Исходя из логики что у каждого класса свой диапазон цен

In [211]:
df_first =df.loc[df.Pclass ==1]
df_first

Survived  Pclass  \
PassengerId                     
2                   1       1   
4                   1       1   
7                   0       1   
12                  1       1   
24                  1       1   
28                  0       1   
31                  0       1   
32                  1       1   
35                  0       1   
36                  0       1   
53                  1       1   
55                  0       1   
56                  1       1   
62                  1       1   
63                  0       1   
65                  0       1   
84                  0       1   
89                  1       1   
93                  0       1   
97                  0       1   
98                  1       1   
103                 0       1   
111                 0       1   
119                 0       1   
125                 0       1   
137                 1       1   
138                 0       1   
140                 0       1   
152                 1       1   
156                 0       1   
...               ...     ...   
764                 1       1   
766                 1       1   
767                 0       1   
780                 1       1   
782                 1       1   
783                 0       1   
790                 0       1   
794                 0       1   
797                 1       1   
803                 1       1   
807                 0       1   
810                 1       1   
816                 0       1   
821                 1       1   
823                 0       1   
830                 1       1   
836                 1       1   
840                 1       1   
843                 1       1   
850                 1       1   
854                 1       1   
857                 1       1   
858                 1       1   
863                 1       1   
868                 0       1   
872                 1       1   
873                 0       1   
880                 1       1   
888                 1       1   
890                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
7                                      McCarthy, Mr. Timothy J    male  54.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
24                                Sloper, Mr. William Thompson    male  28.0   
28                              Fortune, Mr. Charles Alexander    male  19.0   
31                                    Uruchurtu, Don. Manuel E    male  40.0   
32              Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN   
35                                     Meyer, Mr. Edgar Joseph    male  28.0   
36                              Holverson, Mr. Alexander Oskar    male  42.0   
53                    Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0   
55                              Ostby, Mr. Engelhart Cornelius    male  65.0   
56                                           Woolner, Mr. Hugh    male   NaN   
62                                         Icard, Miss. Amelie  female  38.0   
63                                 Harris, Mr. Henry Birkhardt    male  45.0   
65                                       Stewart, Mr. Albert A    male   NaN   
84                                     Carrau, Mr. Francisco M    male  28.0   
89                                  Fortune, Miss. Mabel Helen  female  23.0   
93                                 Chaffee, Mr. Herbert Fuller    male  46.0   
97                                   Goldschmidt, Mr. George B    male  71.0   
98                             Greenfield, Mr. William Bertram    male  23.0   
103                                  White, Mr. Richard Frasar    male  21.0   
111          

In [210]:
df_first.Fare.std()

78.380372646728816

In [213]:
df_first.Fare.mean()

84.154687499999994

In [214]:
df_first.Fare.mean()+(df_first.Fare.std()*5)

476.05655073364403

In [215]:
df_first.loc[df_first.Fare >=df_first.Fare.mean()+(df_first.Fare.std()*5)]

Survived  Pclass                                Name     Sex  \
PassengerId                                                                 
259                 1       1                    Ward, Miss. Anna  female   
680                 1       1  Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1              Lesurer, Mr. Gustave J    male   

              Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
PassengerId                                                                 
259          35.0      0      0  PC 17755  512.3292          NaN        C   
680          36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
738          35.0      0      0  PC 17755  512.3292         B101        C   

             Surname  
PassengerId           
259             Ward  
680          Cardeza  
738          Lesurer

In [216]:
df_second =df.loc[df.Pclass ==2]
df_second.loc[df_second.Fare >=df_second.Fare.mean()+(df_second.Fare.std()*5)]

Empty DataFrame
Columns: [Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked, Surname]
Index: []

In [217]:
df_third =df.loc[df.Pclass ==3]
df_third.loc[df_third.Fare >=df_third.Fare.mean()+(df_third.Fare.std()*5)]

Empty DataFrame
Columns: [Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked, Surname]
Index: []

Для всех, без разбора по классам.

In [218]:
df.loc[df.Fare >=df.Fare.mean()+(df.Fare.std()*5)]

Survived  Pclass                                Name     Sex  \
PassengerId                                                                 
259                 1       1                    Ward, Miss. Anna  female   
680                 1       1  Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1              Lesurer, Mr. Gustave J    male   

              Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
PassengerId                                                                 
259          35.0      0      0  PC 17755  512.3292          NaN        C   
680          36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
738          35.0      0      0  PC 17755  512.3292         B101        C   

             Surname  
PassengerId           
259             Ward  
680          Cardeza  
738          Lesurer

## Задание 3.3
Выведите на экран столбцы, в которых бывают пропущенные значения. (Hint: *Здесь может помочь метод* `.any()`*, а также индексация переменной* `df.columns`).

In [219]:
df.columns

Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket',
       'Fare', 'Cabin', 'Embarked', 'Surname'],
      dtype='object')

In [221]:
df.Cabin.isnull().any()

True

In [228]:
df.loc[ 0:,'Cabin'].isnull().any()

True

In [231]:
list_col=df.columns
list_col_with_null=[]
for i in list_col:
    if df.loc[ 0:,i].isnull().any()==True:
        list_col_with_null.append(i)
print(list_col_with_null)

['Age', 'Cabin', 'Embarked']
